### PR TITLE
Implement `ExperimentalAttribute`

### LINKED ISSUE #77869: Enable Library-Level Preview Features
***Edited by @terrajobst**. Original request at the end. The spec for this feature is [here][design].*

### Background and motivation

In .NET 6, we've added a way to [ship preview features][preview-features] in otherwise stable releases. It works by tagging APIs (and language features) as preview features and requiring the user's project to explicitly opt-into using them via the `<EnablePreviewFeatures>` property.

The feature was primarily designed to accommodate platform-level preview features, that is for features that span runtime, library, and language, such as generic math. However, we also believed that the feature would work for libraries that want to offer preview functionality in otherwise stable NuGet packages.

This feature is about providing a model that works great for library authors. For details, see the [design document][design].

[preview-features]: ../../2021/preview-features/preview-features.md
[design]: https://github.com/dotnet/designs/blob/main/accepted/2023/preview-apis/preview-apis.md

### API Proposal

```C#
namespace System.Diagnostics.CodeAnalysis;

[AttributeUsage(AttributeTargets.Assembly |
                AttributeTargets.Module |
                AttributeTargets.Class |
                AttributeTargets.Struct |
                AttributeTargets.Enum |
                AttributeTargets.Constructor |
                AttributeTargets.Method |
                AttributeTargets.Property |
                AttributeTargets.Field |
                AttributeTargets.Event |
                AttributeTargets.Interface |
                AttributeTargets.Delegate, Inherited = false)]
public sealed class ExperimentalAttribute : Attribute
{
    public ExperimentalAttribute(string diagnosticId);
    public string DiagnosticId { get; }
    public string? UrlFormat { get; set; }
}
```

#### Shipping an experimental API in a stable package

Martin builds a library that integrates with ASP.NET Core and provides features to make it easier to build services. The library is distributed as a NuGet package. The package is already in a version well past 1.0 and is considered stable.

He ships the package often and doesn't want to maintain multiple flavors of the package, one marked as prerelease and one marked as stable. However, he wants to expose new APIs without making the promise that these APIs are stable. He designs the new APIs as if they would be stable, that is, he puts them in the namespace they belong and names the types and members without any naming convention that implies that they aren't stable yet. If customer feedback is positive, the API will become stable as-is, otherwise he'll make changes to address the feedback.

In order to convey to the consumers that these APIs aren't stable yet, he puts the `Experimental` attribute on them:

```C#
namespace MartinsLibrary.FancyLogging
{
    public static class FancyLoggerExtensions
    {
        [Experimental("ML123", UrlFormat="https://martinslibrary.net/diagnostics/{0}")]
        public static IServiceCollection AddFancyLogging(this IServiceCollections services)
        {
            // ...
        }
    }
}
```

#### Consuming an experimental API

David builds a new version of Jabber.NET and consumes Martin's library. He very much wants to play with the new fancy logging Martin added. When David calls `AddFancyLogging` he gets a compilation error:

> error ML123: FancyLoggerExtensions.AddFancyLogging() is not a stable API. In order to consume this experimental API, suppress this error from the call site.

David is happy to accommodate future breaking changes around the fancy logging feature. He understands that the diagnostic ID is specific to the fancy logging feature of Martin's library, so he decides to add a suppression to the project file:

```xml
<Project>
    <PropertyGroup>
        <!-- Fancy logging is an experimental APIs -->
        <NoWarn>$(NoWarn);ML123</NoWarn>
    </PropertyGroup>
</Project>
```

<details>
<summary>Original Request</summary>
### Background and motivation

Library developers often want to experiment with APIs before committing to long-term support. This is true within an organization, as well as within the open-source community. It would be very useful for the .NET infrastructure to support experimental interfaces in a first class way.

The idea is to flag the use of experimental APIs in calling code, to make developers aware that they're calling into an API which is not yet settled and is subject to change outside of the normal support model for the component.

The attribute would be supplemented by a Roslyn analyzer that reports use of experimental APIs. You can also imagine special treatment in IntelliSense and in DocFx for experimental APIs.

### API Proposal

```csharp
namespace System.Diagnostics.CodeAnalysis;

/// <summary>
/// Indicates that an API element is experimental and subject to change without notice.
/// </summary>
[AttributeUsage(
    AttributeTargets.Class |
    AttributeTargets.Struct |
    AttributeTargets.Enum |
    AttributeTargets.Interface |
    AttributeTargets.Delegate |
    AttributeTargets.Method |
    AttributeTargets.Constructor |
    AttributeTargets.Property |
    AttributeTargets.Field |
    AttributeTargets.Event |
    AttributeTargets.Assembly)]
public sealed class ExperimentalAttribute : Attribute
{
    /// <summary>
    /// Gets a human readable explanation for marking API as experimental.
    /// </summary>
    public string? Message { get; }

    /// <summary>
    /// Initializes a new instance of the <see cref="ExperimentalAttribute"/> class.
    /// </summary>
    public ExperimentalAttribute()
    {
        // Intentionally left empty.
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ExperimentalAttribute"/> class.
    /// </summary>
    /// <param name="message">Human readable explanation for marking experimental API.</param>
    public ExperimentalAttribute(string message)
    {
        Message = Throws.IfNull(message);
    }
}
```

### API Usage

```csharp
[Experimental("Trying this out, maybe committed in the 3.1 release")]
public void DoSomethingNew();

[assembly: Experimental("This whole assembly is subject to change without notice")]
```

### Alternative Designs

_No response_

### Risks

_No response_
</details>

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #77869

/cc: @geeknoid @joperezr @stephentoub

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml
Comment (RussKie) [inline src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml:146]: I _think_ that the attribute is no longer relevant at the point an application is being trimmed. But I don't claim I understand this area too well.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (RussKie) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: I haven't found any docs whether this file is modified manually or by some tool... So I manually edited it.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (RussKie) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:261]: 🤔  I'm curious why we need to explicitly declare each file and can't use globs.

**Comment by RussKie** [comment]
Build failures don't seem related to my change, as other PRs fail on the same CI legs...

File: src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.Shared.xml:146]: Yes, that's fine.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:261]: Because there's a ton of special-casing about which files to include in which build, and in general we care more about explicitness and accuracy than saving a line.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:4]: We've generally not been using file-scoped namespaces. While a few may have snuck in, let's not use it here.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:33]:
```suggestion
```
We generally don't want to throw exceptions from attribute ctors. It makes reflection challenging.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (stephentoub) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: There is a command to do it, but manual editing is ok. It just increases the liklihood of more diffs the next time someone runs the tool.

File: src/libraries/System.Runtime/tests/System/Diagnostics/CodeAnalysis/ExperimentalAttributeTests.cs
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Diagnostics/CodeAnalysis/ExperimentalAttributeTests.cs:20]: Remove

File: src/libraries/System.Runtime/tests/System/Diagnostics/CodeAnalysis/ExperimentalAttributeTests.cs
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/Diagnostics/CodeAnalysis/ExperimentalAttributeTests.cs:29]: Remove.

Whatever tooling is validating these attributes should also issue diagnostics for incorrect usage.

**Comment by stephentoub** [comment]
> Resolves https://github.com/dotnet/runtime/issues/77869

If this resolves that issue, what is tracking the tooling side of this, either in am analyzer or the compiler (wherever we landed).

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (RussKie) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: Can you please point me to the doc describing the command?

**Comment by RussKie** [comment]
> If this resolves that issue, what is tracking the tooling side of this, either in an analyzer or the compiler (wherever we landed)?
> 
> And when do we expect that tooling to land? We don't want to ship the attribute without associated tooling.

@terrajobst is working on that AFAIK.
/cc: @jaredpar

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:24]: @terrajobst, @bartonjs, did we discuss Inherited false/true for this?  Seems like true might actually make more sense; if you derive from something that's experimental, seems you'd also need to be experimental?

Comment (geeknoid) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:24]: The rule we've been using for our version says that:

* You can mark a whole assembly as experimental and that affects every symbol exposed by the assembly.
* You can mark a base class as experimental and that affects every derived class.
* You can mark an outer type as experimental and that affects nested types.
* You can mark a virtual method as experimental and that affects every implementation of the method.

We decided not to do this for interface types however. If your type implements an experimental interface, it does not make your new type experimental. Although I think a point can be made that experimental annotations on interface types also should contaminate implementations.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (joperezr) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: https://github.com/dotnet/runtime/blob/bd7e1cb81d763589b95f7bfbaa2a96dd364e9103/docs/coding-guidelines/updating-ref-source.md

Comment (joperezr) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: Note that this case we are adding it to System.Runtime, so you will need to follow this section: https://github.com/dotnet/runtime/blob/bd7e1cb81d763589b95f7bfbaa2a96dd364e9103/docs/coding-guidelines/updating-ref-source.md#for-systemruntime

**Comment by joperezr** [comment]
@terrajobst or @stephentoub, can you remind me what is the plan for using this attribute downlevel? Is the idea to just define the attribute as part of your assembly as an internal type or were we planning on having a Microsoft.BCL.* package with this?

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (RussKie) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:8285]: Awesome, thank you!

**Comment by stephentoub** [comment]
> can you remind me what is the plan for using this attribute downlevel?

@terrajobst can correct me if I'm wrong, but I believe the plan is:
- We ship a public ExperimentalAttribute in-box in .NET 8.
- We ship either an analyzer / compiler feature in .NET 8 / C# 12 that recognizes any definition of ExperimentalAttribute regardless of visibility.
- Anyone who needs this downlevel can add an internal copy of the attribute to their own assembly.

Immo?

**Review by stephentoub — APPROVED** [review]

**Comment by RussKie** [comment]
@stephentoub the build is green. Is it ok to merge?

**Comment by stephentoub** [comment]
@jaredpar, @terrajobst, when do we expect the tooling to land?

**Comment by stephentoub** [comment]
@RussKie, let's give folks a few more days to comment on the tooling.  I just want to make sure there's a plan in place for that to land in this release before merging the attribute; otherwise, we'll end up ripping it out.

**Comment by RussKie** [comment]
No worries, I'll let you all handle the merge :)

**Comment by tommcdon** [comment]
> @jaredpar, @terrajobst, when do we expect the tooling to land?

@jaredpar @terrajobst would you mind providing an update?

**Comment by joperezr** [comment]
@stephentoub is there anything in particular we are waiting for on the tooling design that we want to be defined before we merge this in? Or putting it in a different way, do we expect anything in this PR to have to change depending on the design of the tooling side? If the answer is no, would it be okay to go ahead and merge this as is then?

**Comment by stephentoub** [comment]
> is there anything in particular we are waiting for on the tooling design that we want to be defined before we merge this in?

Yes, that it's actually going to happen.

**Comment by geeknoid** [comment]
Note that in lieu of doing this in the compiler, an analyzer is sufficient for the 99% case. I happen to have one: https://github.com/dotnet/extensions/blob/main/src/Analyzers/Microsoft.Extensions.ExtraAnalyzers/Common/UsingExperimentalApiAnalyzer.cs. I could migrate it to the main NetAnalyzers world if needed.

**Comment by jaredpar** [comment]
> would you mind providing an update?

Can't promise this will make it to the compiler in .NET 8 right now. It's an issue of bandwidth. Hopeful but can't 100% guarantee.

**Comment by terrajobst** [comment]
> Can't promise this will make it to the compiler in .NET 8 right now. It's an issue of bandwidth. Hopeful but can't 100% guarantee.

Would it make sense to have a fallback plan via an analyzer then? Waiting for a year would be very much undesirable.

**Comment by terrajobst** [comment]
The biggest UX difference between doing it in the compiler and doing it as analzyer is that only the compiler supports an open ended ID set. We can't support a string-based diagnostic ID in the attribute as analyzer has to report which diagnostic IDs it will produce.

But Jared mentioned that it's just difficult to support the same semantics between obsolete and an analyzer but I think we can sufficient fidelity there, albeit with a bug tail as always.

**Comment by geeknoid** [comment]
I'm with Jose here, I don't value the custom diag IDs and after a year or two of using an experimental attribute + analyzer in our code base, this has just not been an issue.

**Comment by joperezr** [comment]
Ok, I've re-watched the [API Proposal](https://www.youtube.com/watch?v=-MBZGdnYPGA) meeting where this was discussed and understand the scenario now. The answer to my question above seems to be that the 1% case is for library authors that want to use a custom diagnostic ID for their experimental APIs. So for example, I build the FooBar library and want to introduce an experimental API, and I want to specify that people calling this API should get a diagnostic warning `Foo0001` when using that API. This way, I may group all of my experimental APIs to all use the same daignostic (`Foo0001`) so that consumers of my library that are okay with using my experimental APIs can simply add a `<NoWarn>Foo0001</NoWarn>` to their project in order to remove all of the warnings coming from uses of my experimental APIs, but not the ones coming from other experimental APIs in different libraries.

That scenario does seem niche, and seems like it was mostly inspired by the [Obsolete attribute](https://apisof.net/catalog/c6dd86d3-f1e8-5e40-69c5-ea35cd72a9ee), which has the same concept. I propose we do an incremental approach here where we start with an attribute that doesn't support custom diagnostic IDs and can be merged now and we can have a temporary analyzer on the roslyn-analyzers repo for this as @geeknoid suggests, and then if we think it's necessary and @jaredpar thinks we have time to get to this in .NET 8, we later re-add the DiagnosticId support and move the analyzer to be part of the compiler instead. Does that sound reasonable?

**Comment by RedworkDE** [comment]
(I only have limited experience with this and didn't try this idea) While analyzers have to declare the diagnostics they want to emit, source generators do not. So shouldn't it be possible to implement this as a very weird source generator, that doesn't emit any sources and just raises the experimental diagnostics with the custom ids?

**Comment by jaredpar** [comment]
I'll give you a firm yes / no by end of week on this.

**Comment by jaredpar** [comment]
We're going to try and get our end done this week / early next. The change itself is small but the testing is not. @jcouv will be doing the work.

**Comment by jaredpar** [comment]
We can start the work without this PR.

**Comment by stephentoub** [comment]
If the tooling is going to happen soon, whether in compiler or analyzer as a backup, I have no problem merging the attribute now.  I just wanted to make sure it was still in the works, as I hadn't heard any update for literally months :)

**Comment by joperezr** [comment]
Sounds good to me. Merging this in then, thanks folks!

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: @joperezr @stephentoub Let's not include "Assembly" and "Module" in the list. I don't think those targets are valid for other attributes of this kind (`Obsolete`, `Windows.Foundation.Metadata.Experimental`, `
Windows.Foundation.Metadata.Deprecated`, ...).

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (jcouv) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:25]: @joperezr @stephentoub I just want to confirm the expected user experience with an example. I'm re-using a message we already have for APIs marked as experimental.

Let me know if you have any feedback.

```
    [Fact]
    public void Test()
    {
        var src = """
C.M();

[System.Diagnostics.CodeAnalysis.Experimental("DiagID1", UrlFormat = "https://example.org/{0}")]
class C
{
    public static void M() { }
}
""";

        var comp = CreateCompilation(new[] { src, experimentalAttributeSrc });
        comp.VerifyDiagnostics(
            // 0.cs(1,1): warning DiagID1: 'C' is for evaluation purposes only and is subject to change or removal in future updates. (https://example.org/DiagID1)
            // C.M();
            Diagnostic("DiagID1", "C").WithArguments("C").WithLocation(1, 1)
            );

        var diag = comp.GetDiagnostics().Single();
        Assert.Equal("DiagID1", diag.Id);
        Assert.Equal(ErrorCode.WRN_Experimental, (ErrorCode)diag.Code);
        Assert.Equal("https://example.org/DiagID1", diag.Descriptor.HelpLinkUri);
    }
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: cc: @terrajobst @geeknoid 

Thanks for the feedback @jcouv. Aside from the fact that other attributes don't support this, is there a particular reason why you don't think we should allow Assembly or Module use? Seems to me that there are valid scenarios where people developing new assemblies might opt to use it at that level, in the past this has been done mostly via using prerelease package versions but I think that not everybody uses that and there are some scenarios where it might be better to have stable p ackages with assemblies marked as experimental in them.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:25]: Yup, the above looks right to me. You can then also write a compilation which passes in the NoWarn for your diagnostic which should not show it any more, or alternatively suppress at each call site by using pragma.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: The main reason for my question is precedent. The secondary reason is implementation cost (deviating from existing attributes means additional complexity and testing).
I'm not saying it's a bad idea, but that this previous design was worked, I think we'd need a strong reason to revise it (why was it sufficient for those other attributes but not this one).

Comment (joperezr) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: fair enough. @bartonjs @terrajobst @stephentoub Do you know off the top of your head why the Obsolete attribute doesn't support Assembly or Module targets? I would have expected the same arguments should be made for both Obsolete and Experimental, since they are essentially the same attribute but just tracking the two opposing ends of an API lifecycle.

Comment (RussKie) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: I believe R9 has assembly-level applications of the attribute. E.g., https://github.com/dotnet/extensions/blob/d558f8de86adf93044bc5386bd4367b3d509ead3/Directory.Build.targets#L68-L72

Comment (terrajobst) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: For the platform compat attributes we allowed assembly/module and we use it quite a bite.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (terrajobst) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:25]: Same

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs
Comment (terrajobst) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: I'm fine with excluding them for now if we want to match ObsoleteAttribute. We can revisit then we run into limitations.

Comment (RussKie) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: @geeknoid FYI

Comment (jcouv) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/ExperimentalAttribute.cs:13]: FYI, the compiler/roslyn change was merged.

**Pull request merged**
